app/merginapi.cpp

```diff
@@ -42,18 +114,122 @@ void MerginApi::listProjectsReplyFinished()
     if (r->error() == QNetworkReply::NoError)
     {
       QByteArray data = r->readAll();
+      cacheProjectsData(data);
       mMerginProjects = parseProjectsData(data);
     }
     else {
         QString message = QStringLiteral("Network API error: %1(): %2").arg("listProjects", r->errorString());
         qDebug("%s", message.toStdString().c_str());
-        emit networkErrorOccurred( r->errorString() );
+        emit networkErrorOccurred( r->errorString(), "Mergin API error: listProjects" );
     }
 
     r->deleteLater();
     emit listProjectsFinished();
 }
 
+void MerginApi::downloadProjectReplyFinished()
+{
+
+    QNetworkReply* r = qobject_cast<QNetworkReply*>(sender());
+    if (!r) {
+        emit networkErrorOccurred( "No reply received.", "Mergin API error: downloadProject" );
+        qDebug() << r->errorString();
+        mPendingRequests.remove(r->url());
+        return;
+    }
+
+    if (r->error() == QNetworkReply::NoError)
+    {
+        QString projectName("temp");
+        if (mPendingRequests.contains(r->url())) {
```

Comment:
could it happen that mPendingRequests does not contain the url? If not, let's just do `QString projectName = mPendingRequests.value(r->url());`